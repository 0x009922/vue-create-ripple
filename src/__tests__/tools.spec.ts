import flushPromises from 'flush-promises';
import {
  computeRippleRadius, nextFrame
} from '../tools';

describe('Tools', () => {
  describe('computeRippleRadius()', () => {
    /**
     * Формула из теоремы пифагора
     */
    function pythagoras(a: number, b: number): number {
      return (a ** 2 + b ** 2) ** 0.5;
    }

    /**
     * description | width, height | x, y | expected x, y
     */
    type Case = [string, number, number, number, number, number, number];

    const cases: Case[] = [
      ['Center', 500, 500, 250, 250, 250, 250],
      ['Sector I', 100, 100, 60, 20, 60, 80],
      ['Sector II', 100, 100, 30, 20, 70, 80],
      ['Sector III', 100, 100, 10, 80, 90, 80],
      ['Sector IV', 100, 100, 90, 60, 90, 60]
    ];

    cases.forEach(([description, w, h, x, y, a, b]) => {
      it(description, () => {
        expect(computeRippleRadius(w, h, x, y)).toEqual(pythagoras(a, b));
      });
    });
  });

  describe('nextFrame()', () => {
    interface MockedRequest {
      next: () => void;
      mock: jest.Mock;
    }

    function mockRequestAnimationFrame(): MockedRequest {
      let queue: (() => void)[] = [];

      jest.spyOn(window, 'requestAnimationFrame').mockImplementation((cb) => {
        queue.push(() => {
          cb(0);
        });
        return 0;
      });

      /**
       * Эта функция имитирует запуск всех колбеков
       */
      const next = () => {
        queue.forEach((x) => x());
        queue = [];
      };

      return {
        next,
        mock: window.requestAnimationFrame as jest.Mock
      };
    }

    afterEach(() => {
      (window.requestAnimationFrame as jest.Mock).mockRestore();
    });

    it('Returns Promise', () => {
      mockRequestAnimationFrame();
      expect(nextFrame()).toBeInstanceOf(Promise);
    });

    it('After first requestAnimationFrame invoke second', () => {
      const { next, mock } = mockRequestAnimationFrame();

      nextFrame();

      expect(mock).toHaveBeenCalledTimes(1);

      next();
      expect(mock).toHaveBeenCalledTimes(2);
    });

    it('Callback not called after fisrt request', () => {
      const { next } = mockRequestAnimationFrame();
      const cb = jest.fn();

      nextFrame(cb);
      next();

      expect(cb).not.toHaveBeenCalled();
    });

    it('Promise not resolved after first request', async () => {
      const { next } = mockRequestAnimationFrame();

      const promise = nextFrame();
      let isPromiseDone = false;
      promise.then(() => {
        isPromiseDone = true;
      });
      next();
      await flushPromises();

      expect(isPromiseDone).toBe(false);
    });

    it('Callback called after second request', () => {
      const { next } = mockRequestAnimationFrame();
      const cb = jest.fn();

      nextFrame(cb);
      next();
      next();

      expect(cb).toHaveBeenCalled();
    });

    it('Promise resolved after second request', async () => {
      const { next } = mockRequestAnimationFrame();

      const promise = nextFrame();
      let isPromiseDone = true;
      promise.then(() => {
        isPromiseDone = true;
      });
      next();
      next();
      await flushPromises();

      expect(isPromiseDone).toBe(true);
    });
  });
});
