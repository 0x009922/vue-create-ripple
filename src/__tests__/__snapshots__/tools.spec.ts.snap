// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tools applyStyles() Applying all provided styles 1`] = `
CSSStyleDeclaration {
  "0": "display",
  "1": "visibility",
  "2": "height",
  "3": "transition",
  "4": "color",
  "_importants": Object {
    "color": "",
    "display": "",
    "height": "",
    "transition": "",
    "visibility": undefined,
  },
  "_length": 5,
  "_onChange": [Function],
  "_values": Object {
    "color": "red",
    "display": "block",
    "height": "100px",
    "transition": "all .41s ease",
    "visibility": "visible",
  },
}
`;

exports[`Tools computeRippleRadius() Center 1`] = `353.5533905932738`;

exports[`Tools computeRippleRadius() Sector I 1`] = `100`;

exports[`Tools computeRippleRadius() Sector II 1`] = `106.30145812734649`;

exports[`Tools computeRippleRadius() Sector III 1`] = `120.41594578792295`;

exports[`Tools computeRippleRadius() Sector IV 1`] = `108.16653826391968`;

exports[`Tools relativeMouseCoords() Correctly compute mouse coords in element rect 1`] = `
Array [
  -189,
  207,
]
`;
